src/backend/distributed/metadata/metadata_utility.c

```diff
@@ -1141,6 +1142,24 @@ LoadShardIntervalList(Oid relationId)
 }
 
 
+/*
+ * LoadShardIntervalListWithRetry reloads the shardintervallist in case a concurrent
+ * shard modification happened.
+ */
+List *
+LoadShardIntervalListWithRetry(Oid relationId)
+{
+	List *shardList = LoadShardIntervalList(relationId);
+
+	/* prevent concurrent placement changes and non-commutative DML statements */
+	LockShardListMetadata(shardList, ShareLock);
```

Comment:
we probably want to reflect the fact that we lock the shards for write in the name

(e.g. LoadShardListForWrites)